File: task.php
<?php

function countConsecutiveDuplicates(array $arr): int
{
    $count = 0;
    $lastNumber = null;
    foreach ($arr as $number) {
        if ($number == $lastNumber) {
            $count++;
        }
        $lastNumber = $number;
    }

    return $count;
}

function generateIntArray(int $length = 100, int $min = -10, int $max = 10 ): array
{
    // Использование array_map с range/array_fill снижает производительность
    $array = [];
    for ($i = 0; $i < $length; $i++) {
        $array[] = rand($min, $max);
    }
    return $array;
}

$array = generateIntArray();
$pairsCount = countConsecutiveDuplicates($array);

echo "Количество последовательных пар одинаковых элементов: $pairsCount" . PHP_EOL;

/* 
    length - 10^7
    generateIntArray time               0.332157 секунд
    countConsecutiveDuplicates time     0.062084 секунд
    ---------------------------------------------------
    length - 10^8
    generateIntArray time               3.682004 секунд
    countConsecutiveDuplicates time     0.657786 секунд
*/

File: task.go
package main

import (
	"fmt"
	"math/rand"
)

func countConsecutiveDuplicates(arr []int) int {
	count := 0
	lastNumber := 0
	hasLastNumber := false

	for _, number := range arr {
		if hasLastNumber && lastNumber == number {
			count++
		}
		lastNumber = number
		hasLastNumber = true
	}

	return count
}

func generateIntArray(length int, min int, max int) []int {
	arr := make([]int, length)
	for i := range length {
		arr[i] = rand.Intn(max-min+1) + min
	}
	return arr
}

func main() {
	array := generateIntArray(100, -10, 10)
	pairsCount := countConsecutiveDuplicates(array)

	fmt.Printf("Количество последовательных пар одинаковых элементов: %d\n", pairsCount)

	/* 
		length - 10^7
		generateIntArray time               0.115442 секунд
		countConsecutiveDuplicates time     0.007288 секунд
		---------------------------------------------------
		length - 10^8
		generateIntArray time               1.145516 секунд
		countConsecutiveDuplicates time     0.074628 секунд
	*/
}
